File: Rosella/src/main/java/me/hydos/rosella/render/vertex/VertexFormat.java
package me.hydos.rosella.render.vertex;

import org.lwjgl.system.MemoryStack;
import org.lwjgl.vulkan.VK10;
import org.lwjgl.vulkan.VkPipelineVertexInputStateCreateInfo;
import org.lwjgl.vulkan.VkVertexInputAttributeDescription;
import org.lwjgl.vulkan.VkVertexInputBindingDescription;

public class VertexFormat {
    private final VertexFormatElement[] elements;
    private final VkVertexInputAttributeDescription.Buffer vkAttributes;
    private final VkVertexInputBindingDescription.Buffer vkBindings;
    private final int size;

    VertexFormat(VertexFormatElement... elements) {
        this.elements = elements;

        int correctedLength = 0;
        for (VertexFormatElement vertexFormatElement : elements) {
            if (vertexFormatElement.vkType() != VertexFormatElements.VK_FORMAT_PADDING) {
                correctedLength++;
            }
        }

        this.vkAttributes = VkVertexInputAttributeDescription.callocStack(correctedLength);

        int offset = 0;
        int elementIdx = 0;
        for (VertexFormatElement element : elements) {
            if (element.vkType() != VertexFormatElements.VK_FORMAT_PADDING) {
                vkAttributes.get(elementIdx)
                        .binding(0)
                        .location(elementIdx)
                        .format(element.vkType())
                        .offset(offset);
                elementIdx++;
            }
            offset += element.byteLength();
        }
        vkAttributes.rewind();
        
        this.size = offset;

        this.vkBindings = VkVertexInputBindingDescription.callocStack(1)
                .binding(0)
                .stride(size)
                .inputRate(VK10.VK_VERTEX_INPUT_RATE_VERTEX);
    }

    public VkVertexInputAttributeDescription.Buffer getVkAttributes() {
        return vkAttributes;
    }

    public VkVertexInputBindingDescription.Buffer getVkBindings() {
        return vkBindings;
    }

    public int getSize() {
        return size;
    }

    public VertexFormatElement[] getElements() {
        return elements;
    }

    public VkPipelineVertexInputStateCreateInfo getPipelineVertexInputStateCreateInfo() {
        MemoryStack stack = MemoryStack.stackGet();
        return VkPipelineVertexInputStateCreateInfo.callocStack(stack)
                .sType(VK10.VK_STRUCTURE_TYPE_PIPELINE_VERTEX_INPUT_STATE_CREATE_INFO)
                .pVertexBindingDescriptions(getVkBindings())
                .pVertexAttributeDescriptions(getVkAttributes());
    }
}

File: Rosella/src/main/kotlin/me/hydos/rosella/render/vertex/VertexFormat.kt
package me.hydos.rosella.render.vertex

import org.lwjgl.vulkan.VK10
import org.lwjgl.vulkan.VkVertexInputAttributeDescription
import org.lwjgl.vulkan.VkVertexInputBindingDescription
import java.util.function.Consumer

/**
 * Used to define the attributes going into a shader.
 */
class VertexFormat(private val elementMap: Map<Int, Element>) {

	private val elements = elementMap.values

	val vkAttributes: VkVertexInputAttributeDescription.Buffer
		get() {
			val attributeDescriptions = VkVertexInputAttributeDescription.callocStack(elementMap.size)

			var offset = 0
			for (entry in elementMap) {
				attributeDescriptions[entry.key]
					.binding(0)
					.location(entry.key)
					.format(entry.value.vkType)
					.offset(offset)
				offset += entry.value.byteLength
			}
			return attributeDescriptions.rewind()
		}

	val vkBindings: VkVertexInputBindingDescription.Buffer
		get() {
			return VkVertexInputBindingDescription.callocStack(1)
				.binding(0)
				.stride(getSize())
				.inputRate(VK10.VK_VERTEX_INPUT_RATE_VERTEX)
		}

	fun getSize(): Int {
		var size = 0
		elements.forEach(Consumer {
			size += it.byteLength
		})
		return size
	}

	enum class DataType(val byteLength: Int) {
		FLOAT(Float.SIZE_BYTES),
		UBYTE(Byte.SIZE_BYTES),
		BYTE(Byte.SIZE_BYTES),
		USHORT(Short.SIZE_BYTES),
		SHORT(Short.SIZE_BYTES),
		UINT(Int.SIZE_BYTES),
		INT(Int.SIZE_BYTES);
	}

	enum class Element(val vkType: Int, val byteLength: Int) {
		POSITION(VK10.VK_FORMAT_R32G32B32_SFLOAT, DataType.FLOAT.byteLength * 3),
		NORMAL(VK10.VK_FORMAT_R32G32B32_SFLOAT, DataType.FLOAT.byteLength * 3),
		COLOR(VK10.VK_FORMAT_R32G32B32_SFLOAT, DataType.FLOAT.byteLength * 3),
		UV(VK10.VK_FORMAT_R32G32_SFLOAT, DataType.FLOAT.byteLength * 2),
		PADDING(VK10.VK_FORMAT_R8_SINT, DataType.BYTE.byteLength),
		GENERIC(VK10.VK_FORMAT_R8_SINT, DataType.BYTE.byteLength),
	}
}